lib/frames.js

```diff
@@ -26,17 +42,255 @@ frames.writeFrame = function(frame, stream) {
+  }
+
+  if (!(decodedPayload[0] instanceof DescribedType)) {
+    throw new errors.MalformedPayloadError('Expected DescribedType from AMQP Payload, but received ' + JSON.stringify(decodedPayload[0]));
+  }
+
+  // read frame
+  var channel = sizeAndDoff.readUInt16BE(6); // Bytes 6 & 7 are channel
+  var messageBuffer = payloadBuffer.slice(decodedPayload[1]);
+
+  var described = decodedPayload[0];
+  if (!framesByDescriptor.hasOwnProperty(described.descriptor)) {
+    throw new errors.MalformedPayloadError('Unknown frame descriptor: ' + described.descriptor);
+  }
+
+  var frame = new framesByDescriptor[described.descriptor](described);
+  if (messageBuffer.length) frame.payload = messageBuffer;
+  if (frameType === FrameType.AMQP) {
+    frame.channel = channel;
+  }
+
+  return frame;
+};
+
+function defineFrame(type, definition) {
+  var FrameBase = (type === FrameType.AMQP) ? frames.AMQPFrame : frames.SaslFrame;
+  var Frame = types.defineComposite(FrameBase, definition);
+  framesByDescriptor[Frame.descriptor.code] = Frame;
+  framesByDescriptor[Frame.descriptor.name] = Frame;
+  return Frame;
+}
+
+// AMQP frames
+frames.OpenFrame = defineFrame(FrameType.AMQP, {
+  name: 'open', code: 0x10,
+  fields: [
+    { name: 'containerId', type: 'string', mandatory: true },
+    { name: 'hostname', type: 'string' },
+    { name: 'maxFrameSize', type: 'uint', default: 4294967295 },
+    { name: 'channelMax', type: 'ushort', default: 65535 },
+    { name: 'idleTimeOut', type: 'uint', default: constants.defaultIdleTimeout },
+    { name: 'outgoingLocales', type: 'symbol', multiple: true, default: constants.defaultOutgoingLocales },
+    { name: 'incomingLocales', type: 'symbol', multiple: true, default: constants.defaultIncomingLocales },
+    { name: 'offeredCapabilities', type: 'symbol', multiple: true },
+    { name: 'desiredCapabilities', type: 'symbol', multiple: true },
+    { name: 'properties', type: 'fields', default: {} }
+  ]
+});
+
+frames.BeginFrame = defineFrame(FrameType.AMQP, {
+  name: 'begin', code: 0x11,
+  fields: [
+    { name: 'remoteChannel', type: 'ushort' },
+    { name: 'nextOutgoingId', type: 'uint', mandatory: true },
+    { name: 'incomingWindow', type: 'uint', mandatory: true },
+    { name: 'outgoingWindow', type: 'uint', mandatory: true },
+    { name: 'handleMax', type: 'uint', default:4294967295 },
+    { name: 'offeredCapabilities', type: 'symbol', multiple: true },
+    { name: 'desiredCapabilities', type: 'symbol', multiple: true },
+    { name: 'properties', type: 'fields', default: {} }
+  ]
+});
+
+frames.AttachFrame = defineFrame(FrameType.AMQP, {
+  name: 'attach', code: 0x12,
+  fields: [
+    { name: 'name', type: 'string', mandatory: true },
+    { name: 'handle', type: 'uint', mandatory: true },
+    { name: 'role', type: 'boolean', mandatory: true },
+    { name: 'sndSettleMode', type: 'ubyte', default: 2 },
```

Comment:
for sure, these were just direct parses from the spec. Also of note: if you look below here to `properties` I included a default of `{}` which is _not_ in the spec , but I included it to satisfy our existing tests. Think we should remove that?